rfd/0116-label-expressions.md

```diff
@@ -0,0 +1,394 @@
+---
+authors: Nic Klaassen (nic@goteleport.com)
+state: draft
+---
+
+# RFD 116 - RBAC Label Expressions
+
+## Required Approvers
+
+* Engineering: @rosstimothy || @codingllama || @fspmarshall
+* Security: @reed || @jentfoo
+* Product: @xinding33 || @klizhentas
+
+## What
+
+This RFD proposes an addition to the Teleport Role spec to enable granular
+matching on resource labels.
+
+Roles currently support the following "label matchers" in the `allow` and `deny`
+sections of the spec:
+
+- `node_labels`
+- `app_labels`
+- `cluster_labels`
+- `kubernetes_labels`
+- `db_labels`
+- `db_service_labels`
+- `windows_desktop_labels`
+
+These would be supplemented by new fields supporting predicate expressions that
+can match resource labels:
```

Comment:
Not a flaws of this RFD, but probably worth calling out when this feature is documented...

If you use labels like this to make RBAC decisions, any manipulation of those labels could lead to privilege escalation.  To me this seems most likely with dynamic labels if a customer had an user who was able to manipulate that dynamic command / process.

An additional consideration that may be worth documenting is this usage should be considered carefully.  Considering the prod label example below an administrator may incorrectly assume that `all_except_prod` includes a deny flag that will take precedent.  Or even if a deny precedent is not assumed, there is an increased potential for interacting with another role in a way that prod access is possible.

This is an inherent risk to this feature, and is not something we can easily address (only idea that comes to mind would be in the UI flagging how roles may interact).  If possible when we write the docs it may be good to articulate this feature in needing to consider the pros and cons vs using deny statements (which are safer).

Reply:
> If you use labels like this to make RBAC decisions, any manipulation of those labels could lead to privilege escalation

Yes, but using labels to make RBAC decisions is the whole point of labels and is how Teleport's RBAC works. I don't see how Nic's work on label expressions really changes anything or introduces any more risk than we already have. Am I missing something?

Reply:
I agree, and to be clear I also believe that this feature can be used safely.  I just wanted to highlight that there is some nuance compared to deny statements that need to be considered.  We should try our best to articulate that in our docs.